portal-impl/src/com/liferay/portal/tools/DatabaseLockRunner.java

```diff
@@ -0,0 +1,265 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.portal.tools;
+
+import com.liferay.petra.function.UnsafeRunnable;
+import com.liferay.petra.string.StringBundler;
+import com.liferay.portal.dao.orm.common.SQLTransformer;
+import com.liferay.portal.kernel.dao.db.DB;
+import com.liferay.portal.kernel.dao.db.DBInspector;
+import com.liferay.portal.kernel.dao.db.DBManagerUtil;
+import com.liferay.portal.kernel.dao.jdbc.DataAccess;
+import com.liferay.portal.kernel.log.Log;
+import com.liferay.portal.kernel.log.LogFactoryUtil;
+import com.liferay.portal.kernel.security.SecureRandomUtil;
+import com.liferay.portal.util.PropsValues;
+
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Timestamp;
+
+/**
+ * @author Luis Ortiz
+ */
+public class DatabaseLockRunner {
+
+	public static void runWithLock(
+			String lockKey, UnsafeRunnable<Exception> unsafeRunnable)
+		throws Exception {
+
+		if (!PropsValues.DATABASE_LOCK_MANAGED_STARTUP) {
+			unsafeRunnable.run();
+
+			return;
+		}
+
+		while (true) {
+			long lockId = _getOrCreateLock(lockKey);
+
+			try (Connection connection = DataAccess.getConnection();
```

Comment:
@LuisOrtiz89 please do some proper inline to reduce the Connection getting down to 1. There should be no reason to keep getting a new connection.

Reply:
@shuyangzhou now the class uses just one connection object for all the methods